pkg/liqo-controller-manager/namespace-controller/namespace_controller.go

```diff
@@ -0,0 +1,270 @@
+				klog.Error(err, " --> Unable to List all virtual nodes")
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+
+			if len(nodes.Items) == 0 {
+				klog.Info(" No VirtualNode at the moment")
+				return ctrl.Result{}, nil
+			}
+
+			for _, node := range nodes.Items {
+				i := 0
+				dim := len(node.Labels)
+				id := node.Annotations[liqo_controller_manager.VirtualNodeClusterId]
+				for key := range node.GetLabels() {
+					i++
+					if strings.HasPrefix(key, offloadingPrefixLabel) {
+						if _, ok = namespace.GetLabels()[key]; !ok {
+							klog.Infof(" Namespace '%s' cannot be offloaded on remote cluster: %s", namespace.GetName(), id)
+							break
+						}
+					}
+
+					if i == dim {
+						klog.Infof(" Offload namespace '%s' on remote cluster: %s", namespace.GetName(), id)
+						if err := r.createRemoteNamespace(namespace, remoteNamespaceName, removeMappings[id]); err != nil {
+							return ctrl.Result{}, err
+						}
+						delete(removeMappings, id)
+					}
+
+				}
+			}
+		}
+
+	}
+
+	if len(removeMappings) > 0 {
+		klog.Info(" Delete all unnecessary mapping in NamespaceMaps")
+		if err := r.removeRemoteNamespaces(namespace.GetName(), removeMappings); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func mappingLabelUpdate(oldLabels map[string]string, newLabels map[string]string) bool {
+	ret := false
+	if val1, ok := oldLabels[mappingLabel]; ok {
+		ret = val1 != newLabels[mappingLabel]
+	}
+	return ret
+}
+
+func mappingLabelPresence(labels map[string]string) bool {
+	_, ok := labels[mappingLabel]
+	return ok
+}
+
+// Events not filtered:
+// 1 -- deletion timestamp is updated on a relevant namespace (only that ones with my finalizer)
+// 2 -- add/delete labels, and mappingLabel is present before or after the update
+// 3 -- update the value of mappingLabel label only
+// 4 -- add namespace with at least mappingLabel
+func manageLabelPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !(e.MetaNew.GetDeletionTimestamp().IsZero()) && slice.ContainsString(e.MetaNew.GetFinalizers(), namespaceFinalizer, nil) {
+				return true
+			}
+			oldLabels := e.MetaOld.GetLabels()
+			newLabels := e.MetaNew.GetLabels()
+			return ((len(oldLabels) != len(newLabels)) && (mappingLabelPresence(oldLabels) || mappingLabelPresence(newLabels))) || mappingLabelUpdate(oldLabels, newLabels)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return mappingLabelPresence(e.Meta.GetLabels())
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+}
+
+func (r *NamespaceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Namespace{}).
+		WithEventFilter(manageLabelPredicate()).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 0}). // no concurrent reconcile
```

Comment:
Is it a problem having multiple concurrent reconciles?

Reply:
No just for debugging, i will remove it, thank you